### PR TITLE
Reimplement GC for new storage model + preparation to support non-rel objects

### PR DESCRIPTION
**Pull request by knizhnik**
Based on the results of our discussions, I removed special fork numbers
and introduce ObjectTag/ObjectValue enums which allows to support multiple object types.

Storage API was extended by list_objects() and unlink() methods, on top of which storage independent GC  was implemented.

### CODE REVIEW

**Comment by lubennikovaav** [comment]
@knizhnik , FYI. I've force pushed rebase to master to trigger CI.
Overall this PR looks great. Now it's clear what types of objects we have.
I think this code should have more comments. I am adding them now.

**Comment by hlinnaka** [comment]
Can you split this into two PRs, please?

1. Fix the GC code
2. Add support for non-rel object types

The first is pretty straightforward, but second is a bit hard to review without seeing the rest of the work.

**Comment by knizhnik** [comment]
There is almost no support for non-rel objects in this branch. Just enums value reserved for them.

File: pageserver/src/object_repository.rs [hunk 1]
```diff
@@ -810,56 +783,40 @@ impl ObjectTimeline {
...
-    // TODO: Not used. Currently, we never drop relations. The parsing
-    // of relation drops in COMMIT/ABORT records has not been
-    // implemented. We should also have a mechanism to remove
-    // "orphaned" relfiles, if the compute node crashes before writing
-    // the COMMIT/ABORT record.
+    /// RelationSize. We store it separately not only to ansver nblocks requests faster.
+    /// We also need it to support relation truncation.
+    RelationSize(u32),
+    /// TODO Add a comment
     Unlink,
```
Comment (lubennikovaav) [inline pageserver/src/object_repository.rs:806]: Do I get it right, that to return a relation page at some lsn, we need to check first if the file is not Unlinked, then check RelationSize to ensure that this page is not truncated?

**Review by lubennikovaav — COMMENTED** [review]
ObjectKey, ObjectTag, and ObjectValue definitions seem to be scattered across different modules quite randomly. At least I can't follow the logic.

So I suggest refactoring them later when the remaining non-rel code will be added.

**Comment by hlinnaka** [comment]
> There is almost no support for non-rel objects in this branch. Just enums value reserved for them.

Most of interface changes where BufferTag is replaced with ObjectTag are caused by that. And I can see some special code to handle two-phase objects in the garbage collection code.

At a quick glance I think those BufferTag->ObjectTag changes will need some more thinking. For example, what does it mean to pass a `ObjectTag::RelationMetadata(RelTag)` key to `put_wal_record`? It doesn't make sense to me, but the interface allows it. So I want to review that more closely, and I don't want those discussions to hold up the GC changes. Or vice versa. That's why I'd like to see these as it two separate commits and separate PRs. Makes for a nicer git history too, if we need to look back.

**Comment by knizhnik** [comment]
I created yet another branch gc_only

**Comment by kelvich** [comment]
We discussed this pr with @knizhnik and decided to wait with it until nonrel_wal lands in main

**Comment by kelvich** [comment]
Splitting this into two pr's:
PR for non-rel wal: https://github.com/zenithdb/zenith/pull/227
PR for GC (on hold while non-rel isn't merged): https://github.com/zenithdb/zenith/pull/223

**Pull request closed without merging**